Add a LightCurve.cdpp() method

barentsen opened the pull request:
I made an attempt to implement a `LightCurve.cdpp()` method.  I also found myself adding `LightCurve.remove_outliers()`  and `LightCurve.remove_nans()` as part of the implementation.

This is going to need unit tests and a critical review!

barentsen commented on pyke/lightcurve.py:
```diff
@@ -30,7 +31,10 @@ class LightCurve(object):
+        if flux_err is not None:
+            self.flux_err = np.asarray(flux_err)
+        else:
+            self.flux_err = None
```
I changed this to prevent `flux_error` from being equal to `array(None)`, which breaks code that relies on `flux_err is not None`.

mirca replied:
lgtm

barentsen commented on pyke/lightcurve.py:
```diff
@@ -79,12 +83,14 @@ def flatten(self, window_length=101, polyorder=3, **kwargs):
             Trend in the lightcurve data
         """
-        trend_signal = signal.savgol_filter(x=self.flux, window_length=window_length,
+        lc_clean = self.remove_nans()  # The SG filter does not allow NaNs
```
I added a line to remove NaNs inside `flatten()` because `savgol_filter` seems to break when it is passed NaNs.  Not sure if this is the best approach.

mirca replied:
👍

mirca commented on pyke/lightcurve.py:
```diff
@@ -115,6 +121,107 @@ def fold(self, period, phase=0.):
+        running_mean = np.convolve(cleaned_lc.flux, mean_filter, mode='same')
+        cdpp_ppm = norm_factor * np.std(running_mean) * 1e6
+        return cdpp_ppm
+
```
This looks great!
Should we use running mean or running median? If the later, see
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.medfilt.html

barentsen replied:
The papers mentioned in the docstring use the mean, so that's what we should use.  I guess it makes sense because the median would jump around too much if the window size is small!

mirca replied:
Cool! ``numpy.cumsum`` looks to be faster than ``numpy.convolve``, see stackoverflow: https://stackoverflow.com/questions/13728392/moving-average-or-running-mean

mirca commented:
This looks great! Let's add unit tests!

barentsen commented:
To match Jeff van Cleve's (JVC) implementation, I believe the fluxes need to be gap-filled and mean-divided prior to computing cdpp.  This is still to be added to this implementation.

I also asked JVC the following two questions:

barentsen commented:
I changed the implementation to use the `np.cumsum`-based running mean, for which I have added the `utils.running_mean()` function.

I also added unit tests which verify that `cdpp()` isn't returning nonsense.

It's not 100% clear yet that this is the exact way we should compute cdpp, but I am happy for it to be merged so we can start using it on real data.

mirca commented:
👍
